####### Evaluated the expression of the LHAgenes in Age-Related Diseases (ARD) #########

setwd("Projects/LLIImmune/ARD")
library("GEOquery")
####### Alzheimer's Disease #################
gse <- GEOquery::getGEO("GSE140829", GSEMatrix = TRUE)
fullPhnotype=gse[[1]]@phenoData@data
colnames(fullPhnotype)
phenotype=fullPhnotype[,c("title","age_at_draw:ch1","apoe_status:ch1","batch:ch1","diagnosis:ch1","Sex:ch1")]
colnames(phenotype)=c("sampleID","Age","Apoe","Batch","BioGroup","Sex")
write.table(phenotype,file="AD/sampleInfo.txt",sep="\t",quote=F)

sampleInfo=read.table("/sampleInfo.txt",header=T,row.names=2)
dataExpr=read.table("AD/rawData/GSE140829_final_normalized_data.txt",header=T,row.names=1,check.names=F)
sampleInfo=sampleInfo[order(sampleInfo$BioGroup,sampleInfo$Sex,sampleInfo$Apoe,sampleInfo$Age),]
sampleList=intersect(rownames(sampleInfo),colnames(dataExpr))
length(sampleList) #551 samples
dataExpr=dataExpr[,sampleList]
sampleInfo=sampleInfo[sampleList,]
all(colnames(dataExpr)==rownames(sampleInfo)) #should be TRUE

dataExpr.adjust = limma::removeBatchEffect(dataExpr,batch=sampleInfo$Batch,covariates=model.matrix(~Age+Sex,data=sampleInfo),design = model.matrix(~BioGroup,data=sampleInfo))
write.table(dataExpr.adjust,file="GSEA/ADPBMCExprMatrix.adjust.txt",sep="\t",quote=F)
write.table(sampleInfo,file="GSEA/ADPBMCsampleInfo.txt",sep="\t",quote=F)

table(sampleInfo$BioGroup)
#AD	Control	MCI 
#198	229	124

summary(sampleInfo[sampleInfo$BioGroup=="AD",]$Age)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 61.00   67.25   72.00   72.92   79.00   88.00
summary(sampleInfo[sampleInfo$BioGroup=="MCI",]$Age)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#61.00   69.00   73.00   73.37   78.00   87.00
summary(sampleInfo[sampleInfo$BioGroup=="Control",]$Age)
#Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#61.00   69.00   74.00   73.51   78.00   89.00 

#GSEA input files including ADPBMCExprMatrix.adjust.gct, ADPBMCsampleInfo.cls were manually prepaired dependent on the GSEA/ADPBMCExprMatrix.adjust.txt and GSEA/ADPBMCsampleInfo.txt, respectively.


AD/rawData/GSE140829_final_normalized_data.txt #perl script to integrate the normalzied gene expression form GEO
AD/GSEA/ADPBMCExprMatrix.adjust.gct #gct file (converted from GeneExprIntegrated.txt) for GSEA
AD/GSEA/ADPBMCsampleInfo.cls #cls file (converted from GeneExprIntegrated.txt) for GSEA

####### Type 2 Diabetes (T2D) mellitus #################
gse <- GEOquery::getGEO("GSE156993", GSEMatrix = TRUE)
fullPhnotype=gse[[1]]@phenoData@data
colnames(fullPhnotype)
phenotype=fullPhnotype[,c("title","age_at_draw:ch1","apoe_status:ch1","batch:ch1","diagnosis:ch1","Sex:ch1")]
colnames(phenotype)=c("sampleID","Age","Apoe","Batch","BioGroup","Sex")
write.table(phenotype,file="T2D/sampleInfo.txt",sep="\t",quote=F)

T2D/rawData/GeneExpr.pl #perl script to download the normalzied gene expression form GEO
T2D/rawData/GeneExprIntegrated.pl #perl script to integrate the normalzied gene expression form GEO
T2D/rawData/GeneExprIntegrated.txt #Normalized gene expression matrix generated by T2D/rawData/GeneExprIntegrated.pl
T2D/rawData/GPL570-55999.txt #gene annotation file

T2D/GSEA/T2DGeneExpr.gct #gct file (converted from GeneExprIntegrated.txt) for GSEA
T2D/GSEA/T2DGeneExpr.cls #cls file (converted from GeneExprIntegrated.txt) for GSEA


####### Coronary artery disease (CAD) mellitus #################
gse <- GEOquery::getGEO("GSE202625", GSEMatrix = TRUE)
pheontype=gse[[1]]@phenoData@data
colnames(pheontype)
phenotype=pheontype[,c("title","age:ch1","group:ch1","Sex:ch1")]
colnames(phenotype)=c("Title","Age","BioGroup","Sex")
write.table(phenotype,file="CAD/sampleInfo.txt",sep="\t",quote=F)


CAD/rawData/GeneExpr.txt #normalized gene expression matrix downloaed from GEO
CAD/rawData/GeneExpr.pl #perl script used to convert the transcript expression to gene expression (the sum of the transcripts belonging to the certain gene were calculated)

CAD/GSEA/GeneExpr4CAD.gct #gct file (converted from GeneExprIntegrated.txt) for GSEA
CAD/GSEA/GeneExpr4CAD.cls #cls file (converted from GeneExprIntegrated.txt) for GSEA
